pkg/services/vimmachine.go

```diff
@@ -496,8 +496,17 @@ func (v *VimMachineService) generateOverrideFunc(ctx *context.VIMMachineContext)
 		if vsphereFailureDomain.Spec.Topology.Datastore != "" {
 			vm.Spec.Datastore = vsphereFailureDomain.Spec.Topology.Datastore
 		}
-		if len(vsphereFailureDomain.Spec.Topology.Networks) > 0 {
-			vm.Spec.Network.Devices = overrideNetworkDeviceSpecs(vm.Spec.Network.Devices, vsphereFailureDomain.Spec.Topology.Networks)
+
+		// Some old cases may still use Networks instead of NetworkConfigs, so we should use the old fields
+		if len(vsphereFailureDomain.Spec.Topology.NetworkConfigs) == 0 && len(vsphereFailureDomain.Spec.Topology.Networks) > 0 {
+			vsphereFailureDomain.Spec.Topology.NetworkConfigs = make([]infrav1.FailureDomainNetwork, len(vsphereFailureDomain.Spec.Topology.Networks))
+			for i := range vsphereFailureDomain.Spec.Topology.Networks {
+				vsphereFailureDomain.Spec.Topology.NetworkConfigs[i].NetworkName = vsphereFailureDomain.Spec.Topology.Networks[i]
+			}
+		}
+
+		if len(vsphereFailureDomain.Spec.Topology.NetworkConfigs) > 0 {
+			vm.Spec.Network.Devices = overrideNetworkDeviceSpecs(vm.Spec.Network.Devices, vsphereFailureDomain.Spec.Topology.NetworkConfigs)
```

Comment:
Together with the above, convert if required and use the converted value.
I would suggest to not change the original object so we don't by accident patch this change  back.

```suggestion
		networkConfigs := vsphereFailureDomain.Spec.Topology.NetworkConfigs
		// Convert the deprecated vsphereFailureDomain.Spec.Topology to networkConfigs if necessary.
		if len(networkConfigs) == 0 && len(vsphereFailureDomain.Spec.Topology.Networks) > 0 {
			networkConfigs = make([]infrav1.FailureDomainNetwork, len(vsphereFailureDomain.Spec.Topology.Networks))
			for i := range vsphereFailureDomain.Spec.Topology.Networks {
				networkConfigs[i].NetworkName = vsphereFailureDomain.Spec.Topology.Networks[i]
			}

		}

		if len(networkConfigs) > 0 {
			vm.Spec.Network.Devices = overrideNetworkDeviceSpecs(vm.Spec.Network.Devices, networkConfigs)
```

Reply:
no problem :) will commit the change here and drop the webhook conversion, keeping only the validation of one or other field. Thanks for the review!